src/js/sonifier.js

```diff
@@ -0,0 +1,402 @@
+
+        // Schedule the next voice label, accounting for both the variable-length
+        // delay (the time to play the preceding sonification) and the fixed-length
+        // gap
+
+        // We shouldn't use the gap if this is the first call for a dataset
+        var pause = noGap? delay : delay+gapDuration;
+        if(fluid.textToSpeech.isSupported()) {
+            synth.scheduler.once(pause, function() {
+                that.applier.change("currentlyPlayingData", currentData);
+                textToSpeech.queueSpeech(currentData.label);
+            });
+        }
+    };
+
+    // Recursion function called from floe.chartAuthoring.sonifier.processSonificationQueue
+
+    floe.chartAuthoring.sonifier.playDataAndQueueNext = function(that) {
+
+        var synth = that.model.synth;
+        var sonificationQueue = that.model.sonificationQueue;
+        var data = sonificationQueue.shift();
+
+        // This prevents an error being thrown if a stop has been called but
+        // this function is queued or in the midst of execution
+        var noteDuration = (data !== undefined) ? floe.chartAuthoring.sonifier.getTotalDuration(data.notes.durations) : 0;
+
+        if(sonificationQueue.length > 0) {
+            floe.chartAuthoring.sonifier.processSonificationQueue(noteDuration, false, that);
+        } else {
+            // Schedule to stop the sonification after the last sonification is
+            // played
+            synth.scheduler.once(noteDuration, function() {
+                that.stopSonification();
+            });
+        }
+        if(data !== undefined) {
+            synth.midiNoteSynth.applier.change("inputs.noteSequencer", data.notes);
+            synth.pianoEnvelopeSynth.applier.change("inputs.envelopeSequencer", data.envelope);
+        }
+    };
+
+    // Given an array containing indiividual durations, accumulate them and
+    // return the total duration
+    floe.chartAuthoring.sonifier.getTotalDuration = function (durationsArray) {
+        var sum = function(duration, runningTotal) {
+            return duration+runningTotal;
+        };
+        return fluid.accumulate(durationsArray, sum, 0);
+    };
+
+    floe.chartAuthoring.sonifier.stopSonification = function(that) {
+
+        if(that.model.isPlaying) {
+            try {
+                // Flush the sonification queue
+                that.applier.change("sonificationQueue",[]);
+                // Flush any outstanding flocking schedulers
+                that.model.synth.scheduler.clearAll();
+                // Pause the synth
+                that.model.synth.pause();
+                // Stop any queued voices
+                that.textToSpeech.cancel();
+            } catch(e) {
+
+            } finally {
+                // Always stop the flocking environment
+                flockingEnvironment.stop();
+                // Reset the bus manager
+                flockingEnvironment.busManager.reset();
```

Comment:
Can try and document these issues more thoroughly later, but off the top of my head I found the following when working to implement the stop functionality. It's possible / likely that the issues I encountered were because of problems with the code's approach to using Flocking that you highlighted in comments below:

**Attempt #1 - flockingEnvironment.stop()**

This would cause successive attempts to play the sonification following a stop to fail with a "cannot acquire bus" error.

**Attempt #2 - flockingEnvironment.reset()**

This would cause successive attempts to play the sonification to fail without an obvious reason.

**Current implementation - flockingEnvironment.stop() + flockingEnvironment.busManager.reset()**

This was the one approach I found that would stop the current play while still allowing successive play events.